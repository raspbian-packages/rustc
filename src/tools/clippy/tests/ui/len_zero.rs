#![feature(plugin)]
#![plugin(clippy)]

#![warn(len_without_is_empty, len_zero)]
#![allow(dead_code, unused)]

pub struct PubOne;

impl PubOne {
    pub fn len(self: &Self) -> isize {
        1
    }
}

impl PubOne { // A second impl for this struct - the error span shouldn't mention this
    pub fn irrelevant(self: &Self) -> bool {
        false
    }
}

// Identical to PubOne, but with an allow attribute on the impl complaining len
pub struct PubAllowed;

#[allow(len_without_is_empty)]
impl PubAllowed {
    pub fn len(self: &Self) -> isize {
        1
    }
}

// No allow attribute on this impl block, but that doesn't matter - we only require one on the
// impl containing len.
impl PubAllowed {
    pub fn irrelevant(self: &Self) -> bool {
        false
    }
}

struct NotPubOne;

impl NotPubOne {
    pub fn len(self: &Self) -> isize { // no error, len is pub but `NotPubOne` is not exported anyway
        1
    }
}

struct One;

impl One {
    fn len(self: &Self) -> isize { // no error, len is private, see #1085
        1
    }
}

pub trait PubTraitsToo {
    fn len(self: &Self) -> isize;
}

impl PubTraitsToo for One {
    fn len(self: &Self) -> isize {
        0
    }
}

trait TraitsToo {
    fn len(self: &Self) -> isize; // no error, len is private, see #1085
}

impl TraitsToo for One {
    fn len(self: &Self) -> isize {
        0
    }
}

struct HasPrivateIsEmpty;

impl HasPrivateIsEmpty {
    pub fn len(self: &Self) -> isize {
        1
    }

    fn is_empty(self: &Self) -> bool {
        false
    }
}

pub struct HasIsEmpty;

impl HasIsEmpty {
    pub fn len(self: &Self) -> isize {
        1
    }

    fn is_empty(self: &Self) -> bool {
        false
    }
}

struct Wither;

pub trait WithIsEmpty {
    fn len(self: &Self) -> isize;
    fn is_empty(self: &Self) -> bool;
}

impl WithIsEmpty for Wither {
    fn len(self: &Self) -> isize {
        1
    }

    fn is_empty(self: &Self) -> bool {
        false
    }
}

pub struct HasWrongIsEmpty;

impl HasWrongIsEmpty {
    pub fn len(self: &Self) -> isize {
        1
    }

    pub fn is_empty(self: &Self, x : u32) -> bool {
        false
    }
}

pub trait Empty {
    fn is_empty(&self) -> bool;
}

pub trait InheritingEmpty: Empty { //must not trigger LEN_WITHOUT_IS_EMPTY
    fn len(&self) -> isize;
}



fn main() {
    let x = [1, 2];
    if x.len() == 0 {
        println!("This should not happen!");
    }

    if "".len() == 0 {
    }

    let y = One;
    if y.len()  == 0 { //no error because One does not have .is_empty()
        println!("This should not happen either!");
    }

    let z : &TraitsToo = &y;
    if z.len() > 0 { //no error, because TraitsToo has no .is_empty() method
        println!("Nor should this!");
    }

    let has_is_empty = HasIsEmpty;
    if has_is_empty.len() == 0 {
        println!("Or this!");
    }
    if has_is_empty.len() != 0 {
        println!("Or this!");
    }
    if has_is_empty.len() > 0 {
        println!("Or this!");
    }
    assert!(!has_is_empty.is_empty());

    let with_is_empty: &WithIsEmpty = &Wither;
    if with_is_empty.len() == 0 {
        println!("Or this!");
    }
    assert!(!with_is_empty.is_empty());

    let has_wrong_is_empty = HasWrongIsEmpty;
    if has_wrong_is_empty.len() == 0 { //no error as HasWrongIsEmpty does not have .is_empty()
        println!("Or this!");
    }
}

fn test_slice(b: &[u8]) {
    if b.len() != 0 {
    }
}
