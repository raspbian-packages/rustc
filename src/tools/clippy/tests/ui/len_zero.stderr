error: item `PubOne` has a public `len` method but no corresponding `is_empty` method
  --> $DIR/len_zero.rs:9:1
   |
9  | / impl PubOne {
10 | |     pub fn len(self: &Self) -> isize {
11 | |         1
12 | |     }
13 | | }
   | |_^
   |
   = note: `-D len-without-is-empty` implied by `-D warnings`

error: trait `PubTraitsToo` has a `len` method but no (possibly inherited) `is_empty` method
  --> $DIR/len_zero.rs:55:1
   |
55 | / pub trait PubTraitsToo {
56 | |     fn len(self: &Self) -> isize;
57 | | }
   | |_^

error: item `HasIsEmpty` has a public `len` method but a private `is_empty` method
  --> $DIR/len_zero.rs:89:1
   |
89 | / impl HasIsEmpty {
90 | |     pub fn len(self: &Self) -> isize {
91 | |         1
92 | |     }
...  |
96 | |     }
97 | | }
   | |_^

error: item `HasWrongIsEmpty` has a public `len` method but no corresponding `is_empty` method
   --> $DIR/len_zero.rs:118:1
    |
118 | / impl HasWrongIsEmpty {
119 | |     pub fn len(self: &Self) -> isize {
120 | |         1
121 | |     }
...   |
125 | |     }
126 | | }
    | |_^

error: length comparison to zero
   --> $DIR/len_zero.rs:140:8
    |
140 |     if x.len() == 0 {
    |        ^^^^^^^^^^^^ help: using `is_empty` is more concise: `x.is_empty()`
    |
    = note: `-D len-zero` implied by `-D warnings`

error: length comparison to zero
   --> $DIR/len_zero.rs:144:8
    |
144 |     if "".len() == 0 {
    |        ^^^^^^^^^^^^^ help: using `is_empty` is more concise: `"".is_empty()`

error: length comparison to zero
   --> $DIR/len_zero.rs:158:8
    |
158 |     if has_is_empty.len() == 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is more concise: `has_is_empty.is_empty()`

error: length comparison to zero
   --> $DIR/len_zero.rs:161:8
    |
161 |     if has_is_empty.len() != 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is more concise: `!has_is_empty.is_empty()`

error: length comparison to zero
   --> $DIR/len_zero.rs:164:8
    |
164 |     if has_is_empty.len() > 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is more concise: `!has_is_empty.is_empty()`

error: length comparison to zero
   --> $DIR/len_zero.rs:170:8
    |
170 |     if with_is_empty.len() == 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is more concise: `with_is_empty.is_empty()`

error: length comparison to zero
   --> $DIR/len_zero.rs:182:8
    |
182 |     if b.len() != 0 {
    |        ^^^^^^^^^^^^ help: using `is_empty` is more concise: `!b.is_empty()`

error: trait `DependsOnFoo` has a `len` method but no (possibly inherited) `is_empty` method
   --> $DIR/len_zero.rs:189:1
    |
189 | / pub trait DependsOnFoo: Foo {
190 | |     fn len(&mut self) -> usize;
191 | | }
    | |_^

error: aborting due to 12 previous errors

